include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -604,109 +612,96 @@ struct __parallel_reduce_then_scan_scan_submitter<__sub_group_size, __max_inputs
-                        __sub_group, __unary_op, __binary_op, __sub_group_carry, __in_rng, __out_rng, __start_idx, __n,
-                        __inputs_per_item, __subgroup_start_idx, __sub_group_id, __active_subgroups);
+                        __sub_group, __gen_scan_input, __scan_pred, __reduce_op, __final_op, __sub_group_carry,
+                        __in_rng, __out_rng, __start_idx, __n, __inputs_per_item, __subgroup_start_idx, __sub_group_id,
+                        __active_subgroups);
                 }
+                //if at the last element in the sequence, then we need to write out the last carry out
+                if (__sub_group_local_id == 0 && (__active_groups == __g + 1) &&
+                    (__active_subgroups == __sub_group_id + 1))
+                {
+                    if (__block_num + 1 == __num_blocks)
+                    {
+                        __res_ptr[0] = __sub_group_carry.__v;
+                    }
+                    else
+                    {
+                        //capture the last carry out for the next block
+                        __tmp_ptr[__num_sub_groups_global] = __sub_group_carry.__v;
+                    }
+                }
+
+                __sub_group_carry.__destroy();
             });
         });
     }
 
-    const sycl::nd_range<1> __nd_range;
-
     const std::size_t __max_block_size;
     const std::size_t __num_sub_groups_local;
     const std::size_t __num_sub_groups_global;
     const std::size_t __num_work_items;
+    const std::size_t __num_blocks;
     const std::size_t __n;
 
-    const _BinaryOperation __binary_op;
-    const _UnaryOperation __unary_op;
+    const _GenReduceInput __gen_reduce_input;
+    const _ReduceOp __reduce_op;
+    const _GenScanInput __gen_scan_input;
+    const _ScanPred __scan_pred;
+    const _FinalOp __final_op;
     _InitType __init;
 
     // TODO: Add the mask functors here to generalize for scan-based algorithms
 };
 
-template <typename _ExecutionPolicy, typename _InRng, typename _OutRng, typename _BinaryOperation,
-          typename _UnaryOperation, typename _InitType, typename _Inclusive>
+// General scan-like algorithm helpers
+// _GenReduceInput - a function which accepts the input range and index to generate the data needed by the main output
+//                   used in the reduction operation (to calculate the global carries)
+// _GenScanInput - a function which accepts the input range and index to generate the data needed by the final scan
+//                 and write operations, for scan patterns
+// _ScanPred - a unary function applied to the ouput of `_GenScanInput` to extract the component used in the scan, but
+//             not the part only required for the final write operation
+// _ReduceOp - a binary function which is used in the reduction and scan operations
+// _FinalOp - a function which accepts output range, index, and output of `_GenScanInput` applied to the input range
```

Comment:
I don't mean to bikeshed too much, but perhaps we could consider a different name for the "final op"? The name doesn't really convey what its intended use is. 

From what I can tell, this operator will typically be used for writing the output. Maybe we can consider "write op" or "output op" or maybe even "final write op" if we want to keep the word final.

Reply:
I think its important to make these names good to make the algorithm understandable and usable for other patterns, so please help suggest names.  I agree `_WriteOp` is better, I'll make the change.

Reply:
done